stats/opencensus/stats.go

```diff
@@ -151,47 +151,29 @@ func recordDataBegin(ctx context.Context, b *stats.Begin) {
 // recordDataOutPayload records the length in bytes of outgoing messages and
 // increases total count of sent messages both stored in the RPCs (attempt on
 // client side) context for use in taking measurements at RPC end.
-func recordDataOutPayload(ctx context.Context, op *stats.OutPayload) {
-	d, ok := ctx.Value(rpcDataKey{}).(*rpcData)
-	if !ok {
-		// Shouldn't happen, as gRPC calls tagRPC which populates the rpcData in
-		// context.
-		return
-	}
+func recordDataOutPayload(d *rpcData, op *stats.OutPayload) {
 	atomic.AddInt64(&d.sentMsgs, 1)
 	atomic.AddInt64(&d.sentBytes, int64(op.Length))
 }
 
 // recordDataInPayload records the length in bytes of incoming messages and
 // increases total count of sent messages both stored in the RPCs (attempt on
 // client side) context for use in taking measurements at RPC end.
-func recordDataInPayload(ctx context.Context, ip *stats.InPayload) {
-	d, ok := ctx.Value(rpcDataKey{}).(*rpcData)
-	if !ok {
-		// Shouldn't happen, as gRPC calls TagRPC which populates the rpcData in
-		// context.
-		return
-	}
+func recordDataInPayload(d *rpcData, ip *stats.InPayload) {
 	atomic.AddInt64(&d.recvMsgs, 1)
 	atomic.AddInt64(&d.recvBytes, int64(ip.Length))
 }
 
 // recordDataEnd takes per RPC measurements derived from information derived
 // from the lifetime of the RPC (RPC attempt client side).
-func recordDataEnd(ctx context.Context, e *stats.End) {
-	d, ok := ctx.Value(rpcDataKey{}).(*rpcData)
-	if !ok {
-		// Shouldn't happen, as gRPC calls TagRPC which populates the rpcData in
-		// context.
-		return
-	}
+func recordDataEnd(ctx context.Context, d *rpcData, e *stats.End) {
 	// latency bounds for distribution data (speced millisecond bounds) have
 	// fractions, thus need a float.
 	latency := float64(time.Since(d.startTime)) / float64(time.Millisecond)
 	var st string
 	if e.Error != nil {
-		s, _ := status.FromError(e.Error) // ignore second argument because codes.Unknown is fine
-		st = codes.CodeToStr[s.Code()]
+		s, _ := status.FromError(e.Error)
+		st = internal.CanonicalString.(func(codes.Code) string)(s.Code())
```

Comment:
Please do this cast once globally:

```go
var canonicalString = internal.CanonicalString.(func(codes.Code) string)
```

Reply:
Done. Curious: why do you want this cast globally? I have similar typecasts in the gcp/observability module (GlobalDialOptions/ServerOptions plumbed through internal/), and I use them more than once, but you never mentioned it there.

Reply:
I'm assuming it's so if typecast fails, fails at init time rather than runtime.

Reply:
Yes, that's the big reason.  Also, to avoid doing the cast every time you call it, since it isn't free.